Tests that we can build swan and schism containers and run model configurations with rompy in those containers

benjaminleighton opened the pull request:
This adds comprehensive testing that builds the container and then configures them using rompy and runs the resulting model configuraiton. Additionally the swan container wasn't building, now it builds with a simplified approach, NetCDF and MPI libraries are not compiled packages are used instead, I believe this is likely to be compatible with swan and there is no need for custom compiles but I could be wrong! There are some caveats, existing swan tests are switched off currently due to an issue `@pytest.mark.skip(reason="frequency removed from ModelRun so this template breaks")` to bypass this issue these tests manually patch in frequency to allow a successful swan configuration to be generated. Additionally the schism configuration uses existing test data from switches off station generation. Ongoing work is requried to improve the swan part of this test, it passes now but doesn't yet demonstrate a successfully running swan configuration.

tomdurrant reviewed (commented):
Agree with the swan docker changes. This is a very old dockerfile that predates any of the backend work, and was not intended for this purpose. I agree that a docker for integration testing should be simple. 

I am a little confused by the testing though. It seems to be attempting to re -mplement a fair amount of native functionality, and a bit of odd patching going on. 

Maybe we could have a quick catch up on this... but there are some examples that might be usefule here https://github.com/rom-py/rompy/blob/main/notebooks/backend_examples.ipynb, and the docs here https://rom-py.github.io/rompy/backends.html#using-backend-configurations

Ultimately, the dockerfiles will live in the split repos, so we will need to split this test up to cover the respective cases that we can then add to the respective splits.  The split is essentially done, but its based in scripts, so if you wanted to split into two, I can add those to each case in the split, or alternatively, I can go ahead with what I have, and we can add these to the split repos after they have been split. Or as a middle ground, maybe we could merge the docker changes, and work on the tests from the split repos?

benjaminleighton commented on tests/integration/test_model_containers.py:
```diff
@@ -0,0 +1,559 @@
+
+
+@pytest.fixture(scope="session")
+def schism_image():
```
Removed see comment above

benjaminleighton commented on tests/integration/test_model_containers.py:
```diff
@@ -0,0 +1,559 @@
+
+    backend = DockerRunBackend()
+
+    # Write a helper script to run the exact docker command manually for debugging
```
Removed and I'm now using the backend parameter on ModelRun instead

benjaminleighton commented on tests/integration/test_model_containers.py:
```diff
@@ -0,0 +1,469 @@
+        return False
+
+
+def is_ci_environment() -> bool:
```
It is being used in should_skip_docker_builds which decorates the tests via pytest.mark.skipif. I don't know the best pattern for unit testing slow building docker images though.

benjaminleighton commented on tests/integration/test_model_containers.py:
```diff
@@ -0,0 +1,469 @@
+    grid = SwanGrid(x0=115.68, y0=-32.76, dx=0.001, dy=0.001, nx=50, ny=30, rot=0)
+    
+    # Create a custom ModelRun class that adds frequency to the context
+    class PatchedModelRun(ModelRun):
```
I didn't understand the frequency problem very well but now that I've swapped to SwanConfigComponents as you suggest it works without any special logic. The Swan test also is configured validly so that the model runs and completes without errors in the container. In the older version it was running but failing.

benjaminleighton commented:
Thanks for this additional review @tomdurrant. I hope I've addressed the comments and reduced but haven't yet split the tests. Happy to take your suggestion and merge just the docker file here and I can recreate the remainder of the pull request as two pull requests on schism and swan repos. I'd have split the tests here except I can't figure out an appropriate pattern for shared docker logic between the tests. Is there somewhere we can put common test logic that can be reused from rompy to schism and swan and other repos?

tomdurrant commented:
Ok, for now, I have put the common functions in conftest as fixtures, and split into two. If you are happy with that @benjaminleighton, I'll merge

benjaminleighton commented:
@tomdurrant thanks for the help. I'm happy for it to be merged. Is the plan to then incorporate it into the split or would you prefer me to work this up into new pull requests? Will we duplicate conftest between schism and swan repos for the moment?

tomdurrant commented:
Lets just incorporate as is. conftest is currently replicated across all splits. Lets just get them into the splits so they are not lost. I'll merge as is
